---
title: "Relationship of metabolic syndrome and shoulder OA"
author: "Mark Ziemann"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    fig_width: 7
    fig_height: 7
theme: cosmo
---

Source: https://github.com/markziemann/shoulder-instability-osteroarthritis

# Introduction

What is the relationship between metabolic syndrome (visceral obesity,
hypertension, elevated fasting glucose, and dyslipidaemia) and the
transcriptomic profile of shoulder OA when comparing bone biopsies
between patients with cuff arthropathy and primary osteoarthritis?
For the above analyses to be effective, we will need to adjust for
age, sex and CRP.
Patients are classified as having metabolic syndrome = YES, or not = No,
giving us a discrete variable which takes into account BMI,
Fasting glucose, Cholesterol and Blood pressure in accordance with
the international diabetes federation definition.

```{r,packages}

suppressPackageStartupMessages({
    library(R.utils)
    library("zoo")
    library("tidyverse")
    library("reshape2")
    library("DESeq2")
    library("gplots")
    library("fgsea")
    library("MASS")
    library("mitch")
    library("eulerr")
    library("limma")
    library("topconfects")
    library("kableExtra")
})

```

## Sample sheet

```{r,samplesheet}

ss <- read.table("pheno_data.tsv",header=TRUE)
rownames(ss) <- sapply(strsplit(ss$fastq,"_R1_"),"[[",1)
rownames(ss) <- gsub("_L001","",rownames(ss))
rownames(ss) <- gsub("-",".",rownames(ss))

ss <- ss[order(rownames(ss)),]


```


## Import read counts

Importing osteoarthritis (OA) and shoulder instability (SI) data separately.
Aggregate Lanes together.

```{r,importdata}

if ( !file.exists("counts.rds") ) {
  tmp <- read.table("../fastq/3col.tsv.gz",header=F)
  x <- as.matrix(acast(tmp, V2~V1, value.var="V3", fun.aggregate = sum))
  x <- as.data.frame(x)
  accession <- sapply((strsplit(rownames(x),"\\|")),"[[",2)
  symbol<-sapply((strsplit(rownames(x),"\\|")),"[[",6)
  x$geneid <- paste(accession,symbol)
  xx <- aggregate(. ~ geneid,x,sum)
  rownames(xx) <- xx$geneid
  #colnames <- gsub("T0R","T0",colnames(xx))
  xx$geneid = NULL
  xx <- round(xx)
  txx <- data.frame(t(xx))
  txx$sample <- gsub("_L001","",rownames(txx))
  txx$sample <- gsub("_L002","",txx$sample)
  txx2 <- aggregate(. ~ sample,txx,sum)
  rownames(txx2) <- txx2$sample
  txx2$sample = NULL
  axx <- data.frame(t(txx2))
  axx <- axx[,order(colnames(axx))]
  colnames(axx) <- gsub("^X","",colnames(axx))
  axx <- axx[,order(colnames(axx))]
  remove(tmp,x,xx,txx,txx2)
  gc()
  saveRDS(axx,"counts.rds")
} else {
  axx <- readRDS("counts.rds")
}

```

Runninig some checks to ensure that the sample sheet matches the list of datasets.

```{r,check}

dim(ss)
dim(axx)

duplicated(colnames(axx))

message("sample sheet entries with matching datasets")
length(which(rownames(ss) %in% colnames(axx)))

message("datasets not found in sample sheet")
colnames(axx)[!colnames(axx) %in% rownames(ss)]

```

## QC analysis

Here I'll look at a few different quality control measures.

```{r,qc1,fig.height=7,fig.width=7}

par(mar=c(5,8,3,1))
barplot(colSums(axx),horiz=TRUE,las=1,xlab="OA num reads")
sums <- colSums(axx)
sums <- sums[order(sums)]
barplot(sums,horiz=TRUE,las=1,xlab="OA num reads")
abline(v=15000000,col="red")

```

## MDS plot

Multidimensional scaling plot to show the variation between all samples, very similar to PCA.

```{r,mds,fig.height=7,fig.width=7}

mds <- cmdscale(dist(t(axx)))

plot(mds, xlab="Coordinate 1", ylab="Coordinate 2",
  type = "p",bty="n",pch=19, cex=4 ,col="gray")
text(mds, labels=rownames(mds) )

```

## Metabolic syndrome

Need to:

* select bone datasets only

* look at the effect of metabolic syndrome in OA (ss1).
Rotator cuff (RC) and instability (Insta) dont have bone data.

```{r,dge1}

ss1 <- subset(ss,Tissue=="bone" & Case=="OA")
ss1$Metabolic_Syndrome
ss1$Age
ss1$Sex

x1 <- axx[,colnames(axx) %in% rownames(ss1)]
dim(x1)
x1 <- x1[which(rowMeans(x1)>=10),]
dim(x1)

dds <- DESeqDataSetFromMatrix(countData = x1 , colData = ss1, design = ~ Metabolic_Syndrome )
res <- DESeq(dds)
z<- results(res)
vsd <- vst(dds, blind=FALSE)
zz <- cbind(as.data.frame(z),assay(vsd))
dge <- as.data.frame(zz[order(zz$pvalue),])
dge[1:20,1:6] %>%
  kbl(caption = "Top gene expression differences in bone between OA patients with and without metabolic syndrome.") %>%
  kable_paper("hover", full_width = F)
dge01_basic <- dge

dds <- DESeqDataSetFromMatrix(countData = x1 , colData = ss1, design = ~ Age + Sex + CRP + Metabolic_Syndrome )
res <- DESeq(dds)
z<- results(res)
vsd <- vst(dds, blind=FALSE)
zz <- cbind(as.data.frame(z),assay(vsd))
dge <- as.data.frame(zz[order(zz$pvalue),])
dge[1:20,1:6] %>%
  kbl(caption = "Top gene expression differences in bone between OA patients with and without metabolic syndrome, adjusted for age, sex and CRP.") %>%
  kable_paper("hover", full_width = F)
dge01 <- dge


```

Now look at some plots.
TODO: MDS plot

```{r,deplots1}

maplot <- function(de,contrast_name) {
  sig <-subset(de, padj < 0.05 )
  up <-rownames(subset(de, padj < 0.05 & log2FoldChange > 0))
  dn <-rownames(subset(de, padj < 0.05 & log2FoldChange < 0))
  GENESUP <- length(up)
  GENESDN <- length(dn)
  DET=nrow(de)
  SUBHEADER = paste(GENESUP, "up, ", GENESDN, "down", DET, "detected")
  ns <-subset(de, padj > 0.05 )
  plot(log2(de$baseMean),de$log2FoldChange, 
       xlab="log2 basemean", ylab="log2 foldchange",
       pch=19, cex=0.5, col="dark gray",
       main=contrast_name, cex.main=1)
  points(log2(sig$baseMean),sig$log2FoldChange,
         pch=19, cex=0.5, col="red")
  mtext(SUBHEADER,cex = 1)
}

make_volcano <- function(de,name) {
    sig <- subset(de,padj<0.05)
    N_SIG=nrow(sig)
    N_UP=nrow(subset(sig,log2FoldChange>0))
    N_DN=nrow(subset(sig,log2FoldChange<0))
    DET=nrow(de)
    HEADER=paste(N_SIG,"@5%FDR,", N_UP, "up", N_DN, "dn", DET, "detected")
    plot(de$log2FoldChange,-log10(de$pval),cex=0.5,pch=19,col="darkgray",
        main=name, xlab="log2 FC", ylab="-log10 pval")
    mtext(HEADER)
    grid()
    points(sig$log2FoldChange,-log10(sig$pval),cex=0.5,pch=19,col="red")
}

make_heatmap <- function(de,name,myss,mx,n=30){

  colfunc <- colorRampPalette(c("blue", "white", "red"))
  values <- as.numeric(myss$OA)
  f <- colorRamp(c("yellow", "orange"))
  rr <- range(values)
  svals <- (values-rr[1])/diff(rr)
  colcols <- rgb(f(svals)/255)
  mxn <- mx/rowSums(mx)*1000000
  x <- mxn[which(rownames(mxn) %in% rownames(head(de,n))),]
  heatmap.2(as.matrix(x),trace="none",col=colfunc(25),scale="row", margins = c(10,15), cexRow=0.5, 
    main=paste("Top ranked",n,"genes in",name) , ColSideColors = colcols  )

}

maplot(dge,"insta vs OA")
make_volcano(dge,"insta vs OA")
make_heatmap(dge,"insta vs OA",ss,xx,n=50)

```

## Pathway analysis with mitch

```{r,mitch1}

#download.file("https://reactome.org/download/current/ReactomePathways.gmt.zip", destfile="ReactomePathways.gmt.zip")
#unzip("ReactomePathways.gmt.zip")
genesets <- gmt_import("ReactomePathways.gmt")

# gene table
gt <- as.data.frame(rownames(xx))
gt$gene <- sapply(strsplit(gt[,1]," "),"[[",2)

m <- mitch_import(dge, DEtype="deseq2",geneTable=gt)
m1 <- mitch_calc(m, genesets, priority="effect")
head(m1$enrichment_result,20) %>% 
  kbl(caption = "Top gene pathway differences using the parametric DE data") %>% 
  kable_paper("hover", full_width = F)

m1top <- subset(m1$enrichment_result,p.adjustANOVA<0.05)
m1up <- subset(m1top,s.dist>0)$set
m1dn <- subset(m1top,s.dist<0)$set

mitch_report(m1,outfile="insta_vs_oa_mitchres.html",overwrite=TRUE)

```



## Nonparametric analysis

Looking at gene ranks might be a better option for comparisons across batches.

The rank() function give the rank from lowest to highest and I'll use this for wilcox test.



```{r,rank1}

r <- apply(xx,2,rank)

ctrl <- which(colnames(r) <70)
case <- which(colnames(r) >70)

suppressWarnings(
wtres <- t(apply(r,1,function(x) {
  wt <- wilcox.test(x[case] ,  x[ctrl] )
  meanrankdiff <- mean(x[case]) - mean(x[ctrl])
  unname(c(meanrankdiff,wt$p.value))
})))

colnames(wtres) <- c("meanrankdiff","pvalue")
wtres <- as.data.frame(wtres)
wtres <- wtres[order(wtres$pvalue),]
wtres$fdr <- p.adjust(wtres$pvalue)

top <- head(wtres,50)
top <- top[order(top$meanrankdiff),]

plot(top$meanrankdiff,1:nrow(top),bty="none",xlab="mean rank difference",ylab="gene",pch=19,cex=2,col="gray",main="Top differentially regulated genes")
labels <- sapply(strsplit(rownames(top)," "),"[[",2)
text(top$meanrankdiff,1:nrow(top),labels=labels,cex=0.7)

top

```

## Pathway enrichment

Here I'm using the mitch package and gene pathways from Reactome to understand the affected pathways separately for each tissue type.

Reactome pathways downloaded 9th Dec 2021.

```{r,mitch2}

wtres$stat <- -log10(wtres$pvalue) * sign(wtres$meanrankdiff)
wtres$gene <- sapply(strsplit(rownames(wtres)," "),"[[",2)

wtres2<-wtres[,c("stat","gene")]
wtres2 <- aggregate(. ~ gene, wtres2, max)
rownames(wtres2) <- wtres2$gene
wtres2$gene=NULL

m2 <- mitch_calc(wtres2, genesets, priority="effect")
head(m2$enrichment_result,20) %>% kbl(caption = "Top gene pathway differences using the nonparametric DE data") %>% kable_paper("hover", full_width = F)

m2top <- subset(m2$enrichment_result,p.adjustANOVA<0.05)    
m2up <- subset(m2top,s.dist>0)$set
m2dn <- subset(m2top,s.dist<0)$set

```

## Compare parametric and nonparametric analyses

These analyses gave similar results, so it suggests that there might not be such a huge batch effect impacting the results.

```{r,euler1}


v0 <- list("para up"=m1up,"para dn"=m1dn, "nonp up"=m2up, "nonp dn"=m2dn)

str(v0)

plot(euler(v0),quantities = TRUE, edges = "gray", main="Parametric and nonparametric DE analysis")

```

## Conclusion

Based on the gene expression data, my guess is that there is no major technical batch effect between the two sample groups.
That said, we can't be 100% sure that the differences observed are due to biological differences and not batch effects.
To be certain, it would be a good idea to perform some validation analysis, either by testing the expression of some of these
genes with another method like RTqPCR, or testing pathway activation biochemically.

As OA is thought to associate with inflammation, I was expecting to see these pathways activated.
However OA is different to rheumatoid arthritis which definitely has an autoimmune component to it.

What we did observe was a downregulation of many signaling pathways such as RORA, RAC1, NOTCH2, CREB1, PI3K and others, 
while translation, amino acid recycling, nonsense mediated decay, degradation of mitotic proteins, and p53 pathways were upregulated.

## Session information

For reproducibility.

```{r,sessioninfo}

sessionInfo()

```
