---
title: "Relationship of metabolic syndrome and shoulder OA"
author: "Mark Ziemann"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    fig_width: 7
    fig_height: 7
theme: cosmo
---

Source: https://github.com/markziemann/shoulder-instability-osteroarthritis

# Introduction

What is the relationship between metabolic syndrome (visceral obesity,
hypertension, elevated fasting glucose, and dyslipidaemia) and the
transcriptomic profile of shoulder OA when comparing bone biopsies
between patients with cuff arthropathy and primary osteoarthritis?
For the above analyses to be effective, we will need to adjust for
age, sex and CRP.
Patients are classified as having metabolic syndrome = YES, or not = No,
giving us a discrete variable which takes into account BMI,
Fasting glucose, Cholesterol and Blood pressure in accordance with
the international diabetes federation definition.

Also I will conduct a comparison of cuff arth to primary OA control for age sex and CRP.

## Load libraries and custom plot functions.

Packages.

```{r,packages}

suppressPackageStartupMessages({
    library(R.utils)
    library("zoo")
    library("tidyverse")
    library("reshape2")
    library("DESeq2")
    library("gplots")
    library("fgsea")
    library("MASS")
    library("mitch")
    library("eulerr")
    library("limma")
    library("topconfects")
    library("kableExtra")
})

```

Functions.

```{r,plotfunc}

maplot <- function(de,contrast_name) {
  sig <-subset(de, padj < 0.05 )
  up <-rownames(subset(de, padj < 0.05 & log2FoldChange > 0))
  dn <-rownames(subset(de, padj < 0.05 & log2FoldChange < 0))
  GENESUP <- length(up)
  GENESDN <- length(dn)
  DET=nrow(de)
  SUBHEADER = paste(GENESUP, "up, ", GENESDN, "down", DET, "detected")
  ns <-subset(de, padj > 0.05 )
  plot(log2(de$baseMean),de$log2FoldChange, 
       xlab="log2 basemean", ylab="log2 foldchange",
       pch=19, cex=0.5, col="dark gray",
       main=contrast_name, cex.main=1)
  points(log2(sig$baseMean),sig$log2FoldChange,
         pch=19, cex=0.5, col="red")
  mtext(SUBHEADER,cex = 1)
}

make_volcano <- function(de,name) {
    sig <- subset(de,padj<0.05)
    N_SIG=nrow(sig)
    N_UP=nrow(subset(sig,log2FoldChange>0))
    N_DN=nrow(subset(sig,log2FoldChange<0))
    DET=nrow(de)
    HEADER=paste(N_SIG,"@5%FDR,", N_UP, "up", N_DN, "dn", DET, "detected")
    plot(de$log2FoldChange,-log10(de$pval),cex=0.5,pch=19,col="darkgray",
        main=name, xlab="log2 FC", ylab="-log10 pval")
    mtext(HEADER)
    grid()
    points(sig$log2FoldChange,-log10(sig$pval),cex=0.5,pch=19,col="red")
}

make_heatmap <- function(de,name,groups,mx,n=30){

  colfunc <- colorRampPalette(c("blue", "white", "red"))
  values <- groups
  f <- colorRamp(c("yellow", "orange"))
  rr <- range(values)
  svals <- (values-rr[1])/diff(rr)
  colcols <- rgb(f(svals)/255)
  mxn <- mx/rowSums(mx)*1000000
  x <- mxn[which(rownames(mxn) %in% rownames(head(de,n))),]
  heatmap.2(as.matrix(x),trace="none",col=colfunc(25),scale="row", margins = c(10,15), cexRow=0.5, 
    main=paste("Top ranked",n,"genes in",name) , ColSideColors = colcols  )

}

```

## Sample sheet

```{r,samplesheet}

ss <- read.table("pheno_data.tsv",header=TRUE)
rownames(ss) <- sapply(strsplit(ss$fastq,"_R1_"),"[[",1)
rownames(ss) <- gsub("_L001","",rownames(ss))
rownames(ss) <- gsub("-",".",rownames(ss))

ss <- ss[order(rownames(ss)),]

ss %>%
  kbl(caption = "sample sheet") %>%
  kable_paper("hover", full_width = F)

```

## Import read counts

Importing osteoarthritis (OA) and shoulder instability (SI) data separately.
Aggregate Lanes together.

```{r,importdata}

if ( !file.exists("counts.rds") ) {
  tmp <- read.table("../fastq/3col.tsv.gz",header=F)
  x <- as.matrix(acast(tmp, V2~V1, value.var="V3", fun.aggregate = sum))
  x <- as.data.frame(x)
  accession <- sapply((strsplit(rownames(x),"\\|")),"[[",2)
  symbol<-sapply((strsplit(rownames(x),"\\|")),"[[",6)
  x$geneid <- paste(accession,symbol)
  xx <- aggregate(. ~ geneid,x,sum)
  rownames(xx) <- xx$geneid
  #colnames <- gsub("T0R","T0",colnames(xx))
  xx$geneid = NULL
  xx <- round(xx)
  txx <- data.frame(t(xx))
  txx$sample <- gsub("_L001","",rownames(txx))
  txx$sample <- gsub("_L002","",txx$sample)
  txx2 <- aggregate(. ~ sample,txx,sum)
  rownames(txx2) <- txx2$sample
  txx2$sample = NULL
  axx <- data.frame(t(txx2))
  axx <- axx[,order(colnames(axx))]
  colnames(axx) <- gsub("^X","",colnames(axx))
  axx <- axx[,order(colnames(axx))]
  remove(tmp,x,xx,txx,txx2)
  gc()
  saveRDS(axx,"counts.rds")
} else {
  axx <- readRDS("counts.rds")
}

```

Running some checks to ensure that the sample sheet matches the list of datasets.

```{r,check}

message("Dimesions of sample sheet:")
dim(ss)

message("Dimesions of count matrix:")
dim(axx)

message("Number of duplicated samples in the count matrix:")
length(which(duplicated(colnames(axx))))

message("sample sheet entries with matching datasets")
length(which(rownames(ss) %in% colnames(axx)))

message("datasets not found in sample sheet")
colnames(axx)[!colnames(axx) %in% rownames(ss)]

```

There are some datasets not included in the sample sheet so I will ask Sam to get
those clinical data so they can be included in future.

## QC analysis

Here I'll look at a few different quality control measures.

```{r,qc1,fig.height=7,fig.width=7}

par(mar=c(5,8,3,1))
barplot(colSums(axx),horiz=TRUE,las=1,xlab="OA num reads")
sums <- colSums(axx)
sums <- sums[order(sums)]
barplot(sums,horiz=TRUE,las=1,xlab="OA num reads")
abline(v=15000000,col="red")

```

Some of those read counts are quite low.

## MDS plot

Multidimensional scaling plot to show the variation between all samples, very similar to PCA.

```{r,mds,fig.height=7,fig.width=7}

mds <- cmdscale(dist(t(axx)))

plot(mds, xlab="Coordinate 1", ylab="Coordinate 2",
  type = "p",bty="n",pch=19, cex=4 ,col="gray")
text(mds, labels=rownames(mds) )

```

## Compare cuff arth to primary OA control for age sex and CRP in bone

Need to:

* select Cuff and OA samples

* select bone datasets only

```{r,dge0b}

ss0 <- subset(ss,Cuff_Arthropathy=="Yes" | Primary_OA=="Yes")

ss0b <- subset(ss0,Tissue=="bone")

ss0b %>%
  kbl(caption = "sample sheet for cuff arth vs OA (bone)") %>%
  kable_paper("hover", full_width = F)

message("Metabolic syndrome classification")
ss0b$Metabolic_Syndrome

message("Age data")
ss0b$Age

message("Sex data")
ss0b$Sex

x0b <- axx[,colnames(axx) %in% rownames(ss0b)]
message("count matrix dimensions before filtering out low genes")
dim(x0b)

message("count matrix dimensions after filtering out low genes")
x0b <- x0b[which(rowMeans(x0b)>=10),]
dim(x0b)

x0b <- x0b[,order(colnames(x0b))]
ss0b <- ss0b[order(rownames(ss0b)),]

ss0b %>%
  kbl(caption = "sample sheet") %>%
  kable_paper("hover", full_width = F)

plotMDS(x0b)

message("here is the model spec:")
dds <- DESeqDataSetFromMatrix(countData = x0b , colData = ss0b, design = ~ Primary_OA)
res <- DESeq(dds)
z<- results(res)
vsd <- vst(dds, blind=FALSE)
zz <- cbind(as.data.frame(z),assay(vsd))
dge <- as.data.frame(zz[order(zz$pvalue),])

dge[1:20,1:6] %>%
  kbl(caption = "Top DEGs in bone between cuff arthropathy and OA patients (not correcting for age and sex)") %>%
  kable_paper("hover", full_width = F)

message("here is the model spec:")
dds <- DESeqDataSetFromMatrix(countData = x0b , colData = ss0b, design = ~ Age + Sex + CRP + Primary_OA )
res <- DESeq(dds)
z<- results(res)
vsd <- vst(dds, blind=FALSE)
zz <- cbind(as.data.frame(z),assay(vsd))
dge <- as.data.frame(zz[order(zz$pvalue),])

dge[1:20,1:6] %>%
  kbl(caption = "Top DEGs in bone between cuff arthropathy and OA patients adjusted for age, sex and CRP.") %>%
  kable_paper("hover", full_width = F)

dge0b <- dge

maplot(dge0b,"cuff arthropathy vs OA in bone")
make_volcano(dge0b,"cuff arthropathy vs OA in bone")
make_heatmap(dge0b,"cuff arthropathy vs OA in bone",groups=as.numeric(factor(ss0b$Primary_OA)),x0b,n=50)

```

Now for cuff arthropathy vs OA in capsule.

```{r,dge0c}

ss0 <- subset(ss,Cuff_Arthropathy=="Yes" | Primary_OA=="Yes")

ss0c <- subset(ss0,Tissue=="capsule")

ss0c %>%
  kbl(caption = "sample sheet for cuff arth vs OA (capsule)") %>%
  kable_paper("hover", full_width = F)

message("Metabolic syndrome classification")
ss0c$Metabolic_Syndrome

message("Age data")
ss0c$Age

message("Sex data")
ss0c$Sex

x0c <- axx[,colnames(axx) %in% rownames(ss0c)]
message("count matrix dimensions before filtering out low genes")
dim(x0c)

x0c <- x0c[which(rowMeans(x0c)>=10),]
message("count matrix dimensions after filtering out low genes")
dim(x0c)

x0c <- x0c[,order(colnames(x0c))]
ss0c <- ss0c[order(rownames(ss0c)),]

plotMDS(x0c)

message("here is the model spec:")
dds <- DESeqDataSetFromMatrix(countData = x0c , colData = ss0c, design = ~ Primary_OA)
res <- DESeq(dds)
z<- results(res)
vsd <- vst(dds, blind=FALSE)
zz <- cbind(as.data.frame(z),assay(vsd))
dge <- as.data.frame(zz[order(zz$pvalue),])
dge[1:20,1:6] %>%
  kbl(caption = "Top DEGs in bone between cuff arthropathy and OA patients (not correcting for age and sex)") %>%
  kable_paper("hover", full_width = F)

message("here is the model spec:")
dds <- DESeqDataSetFromMatrix(countData = x0c , colData = ss0c, design = ~ Age + Sex + CRP + Primary_OA )
res <- DESeq(dds)
z<- results(res)
vsd <- vst(dds, blind=FALSE)
zz <- cbind(as.data.frame(z),assay(vsd))
dge <- as.data.frame(zz[order(zz$pvalue),])

dge[1:20,1:6] %>%
  kbl(caption = "Top DEGs in bone between cuff arthropathy and OA patients adjusted for age, sex and CRP.") %>%
  kable_paper("hover", full_width = F)

dge0c <- dge

maplot(dge0c,"cuff arthropathy vs OA in capsule")
make_volcano(dge0c,"cuff arthropathy vs OA in capsule")
make_heatmap(dge0c,"cuff arthropathy vs OA in capsule",groups=as.numeric(factor(ss0b$Primary_OA)),x0b,n=50)

```

## The effect of metabolic syndrome in OA 

Need to:

* select bone datasets only

* select OA samples only

* look at the effect of metabolic syndrome in OA (ss1).

```{r,dge1_b_oa}

ss1 <- subset(ss,Primary_OA=="Yes")

ss1b <- subset(ss1,Tissue=="bone")

message("Metabolic syndrome classification")
ss1b$Metabolic_Syndrome

message("Age data")
ss1b$Age

message("Sex data")
ss1b$Sex

x1b <- axx[,colnames(axx) %in% rownames(ss1b)]
message("count matrix dimensions before filtering out low genes")
dim(x1b)

x1b <- x1b[which(rowMeans(x1b)>=10),]
message("count matrix dimensions after filtering out low genes")
dim(x1b)

x1b <- x1b[,order(colnames(x1b))]
ss1b <- ss1b[order(rownames(ss1b)),]

ss1b %>%
  kbl(caption = "sample sheet") %>%
  kable_paper("hover", full_width = F)

plotMDS(x1b)

message("here is the model spec:")
dds <- DESeqDataSetFromMatrix(countData = x1b , colData = ss1b, design = ~ Metabolic_Syndrome )
res <- DESeq(dds)
z<- results(res)
vsd <- vst(dds, blind=FALSE)
zz <- cbind(as.data.frame(z),assay(vsd))
dge <- as.data.frame(zz[order(zz$pvalue),])

dge[1:20,1:6] %>%
  kbl(caption = "Top DEGs in bone between OA patients with and without metabolic syndrome.") %>%
  kable_paper("hover", full_width = F)

message("here is the model spec:")
dds <- DESeqDataSetFromMatrix(countData = x1b , colData = ss1b, design = ~ Age + Sex + CRP + Metabolic_Syndrome )
res <- DESeq(dds)
z<- results(res)
vsd <- vst(dds, blind=FALSE)
zz <- cbind(as.data.frame(z),assay(vsd))
dge <- as.data.frame(zz[order(zz$pvalue),])

dge[1:20,1:6] %>%
  kbl(caption = "Top DEGs in bone between OA patients with and without metabolic syndrome, adjusted for age, sex and CRP.") %>%
  kable_paper("hover", full_width = F)

dge1b <- dge

```

Now run a mitch analysis.


```{r,mitch1_b_oa}

gnames <- sapply(strsplit(sub("\\."," ",sub("\\."," ",rownames(axx))) ," "),"[[",3)
gt <- data.frame(rownames(axx),gnames)

gs <- gmt_import("ReactomePathways_2023-09-01.gmt")

dge1b$gn <- sapply(strsplit( sub("\\.","_", ( sub("\\.","_",rownames(dge1b)) ) ), "_"),"[[",3)

m <- mitch_import(x=dge1b,DEtype="DESeq2",geneTable=gt)

mres1b <- mitch_calc(x=m,genesets=gs,cores=16,priority="effect")

head(mres1b$enrichment_result,20) %>%
  kbl(caption = "Top Reactomes for dge1b") %>%
  kable_paper("hover", full_width = F)

mitch_report(res=mres1b,outfile="mitch1boa_report.html",overwrite=TRUE)

```

## The effect of metabolic syndrome in cuff arthropathy - consider all tissues

Need to:

* select cuff arthropathy samples only

* look at the effect of metabolic syndrome in OA (ss1).

* consider tissues as "batches" to correct for.

* Unfortunately it isn't possible to correct for participant "batches" at the same time.

```{r,dge1_b_ca}

ss1 <- subset(ss,Cuff_Arthropathy=="Yes")

message("Metabolic syndrome classification")
ss1$Metabolic_Syndrome

message("Age data")
ss1$Age

message("Sex data")
ss1$Sex

x1 <- axx[,colnames(axx) %in% rownames(ss1)]
message("count matrix dimensions before filtering out low genes")
dim(x1)

x1 <- x1[which(rowMeans(x1)>=10),]
message("count matrix dimensions after filtering out low genes")
dim(x1)

x1 <- x1[,order(colnames(x1))]
ss1 <- ss1[order(rownames(ss1)),]

ss1 %>%
  kbl(caption = "sample sheet") %>%
  kable_paper("hover", full_width = F)

plotMDS(x1)

message("here is the model spec:")
dds <- DESeqDataSetFromMatrix(countData = x1 , colData = ss1, design = ~ Tissue + Metabolic_Syndrome )
res <- DESeq(dds)
z<- results(res)
vsd <- vst(dds, blind=FALSE)
zz <- cbind(as.data.frame(z),assay(vsd))
dge <- as.data.frame(zz[order(zz$pvalue),])

dge[1:20,1:6] %>%
  kbl(caption = "Top DEGs that are altered in all tissues by metabolic syndrome.") %>%
  kable_paper("hover", full_width = F)

message("here is the model spec:")
dds <- DESeqDataSetFromMatrix(countData = x1 , colData = ss1, design = ~ Age + Sex + CRP + Metabolic_Syndrome )
res <- DESeq(dds)
z<- results(res)
vsd <- vst(dds, blind=FALSE)
zz <- cbind(as.data.frame(z),assay(vsd))
dge <- as.data.frame(zz[order(zz$pvalue),])

dge[1:20,1:6] %>%
  kbl(caption = "Top DEGs in bone between cuff arthropathy patients with and without metabolic syndrome, adjusted for age, sex and CRP.") %>%
  kable_paper("hover", full_width = F)

dge1 <- dge

```

Now run a mitch analysis.

```{r,mitch1}

gnames <- sapply(strsplit(sub("\\."," ",sub("\\."," ",rownames(axx))) ," "),"[[",3)
gt <- data.frame(rownames(axx),gnames)

gs <- gmt_import("ReactomePathways_2023-09-01.gmt")

dge1$gn <- sapply(strsplit( sub("\\.","_", ( sub("\\.","_",rownames(dge1)) ) ), "_"),"[[",3)

m <- mitch_import(x=dge1,DEtype="DESeq2",geneTable=gt)

mres1 <- mitch_calc(x=m,genesets=gs,cores=16,priority="effect")

head(mres1$enrichment_result,20) %>%
  kbl(caption = "Top Reactomes for dge1b") %>%
  kable_paper("hover", full_width = F)

mitch_report(res=mres1,outfile="mitch1_report.html",overwrite=TRUE)

```

## The effect of metabolic syndrome in rotator cuff

Need to:

* Select rotator cuff

* look at the effect of metabolic syndrome.


```{r,dge2b}

ss2 <- subset(ss,Case=="RC")

ss2c <- subset(ss2,Tissue=="capsule")

message("Metabolic syndrome classification")
ss2c$Metabolic_Syndrome

message("Age data")
ss2c$Age

message("Sex data")
ss2c$Sex

x2c <- axx[,colnames(axx) %in% rownames(ss2c)]
message("count matrix dimensions before filtering out low genes")
dim(x2c)

x2c <- x2c[which(rowMeans(x2c)>=10),]
message("count matrix dimensions after filtering out low genes")
dim(x2c)

x2c <- x2c[,order(colnames(x2c))]
ss2c <- ss2c[order(rownames(ss2c)),]

ss2c <- subset(ss2c,Metabolic_Syndrome=="No" | Metabolic_Syndrome=="Yes")
x2c <- x2c[,colnames(x2c) %in% rownames(ss2c)]

ss2c %>%
  kbl(caption = "sample sheet") %>%
  kable_paper("hover", full_width = F)

plotMDS(x2c)

message("here is the model spec:")
dds <- DESeqDataSetFromMatrix(countData = x2c , colData = ss2c, design = ~ Age + Sex + Metabolic_Syndrome )
res <- DESeq(dds)
z<- results(res)
vsd <- vst(dds, blind=FALSE)
zz <- cbind(as.data.frame(z),assay(vsd))
dge <- as.data.frame(zz[order(zz$pvalue),])

dge[1:20,1:6] %>%
  kbl(caption = "Top DEGs in capsule between RC patients with and without metabolic syndrome.") %>%
  kable_paper("hover", full_width = F)

message("here is the model spec:")
dds <- DESeqDataSetFromMatrix(countData = x2c , colData = ss2c, design = ~ Age + Sex + CRP + Metabolic_Syndrome )
res <- DESeq(dds)
z<- results(res)
vsd <- vst(dds, blind=FALSE)
zz <- cbind(as.data.frame(z),assay(vsd))
dge <- as.data.frame(zz[order(zz$pvalue),])

dge[1:20,1:6] %>%
  kbl(caption = "Top DEGs in bone between OA patients with and without metabolic syndrome, adjusted for age, sex and CRP.") %>%
  kable_paper("hover", full_width = F)

dge2c <- dge

```

## Session information

For reproducibility.

```{r,sessioninfo}

sessionInfo()

```
